pkg/kubelet/util/pluginwatcher/plugin_watcher_test.go

```diff
@@ -0,0 +1,162 @@
+		require.True(t, callbackCount <= (len(testCases)-1))
+		glog.Infof("receives plugin watcher callback for test: %s\n", testCases[callbackCount].description)
+		require.Equal(t, "example-plugin", name, "Plugin name mismatched!!")
+		require.Equal(t, []string{"v1beta1", "v1beta2"}, versions, "Plugin version mismatched!!")
+		// Verifies the grpcServer is ready to serve services.
+		conn, err := grpc.Dial(socketPath, grpc.WithInsecure(), grpc.WithBlock(),
+			grpc.WithTimeout(10*time.Second),
+			grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+				return net.DialTimeout("unix", addr, timeout)
+			}))
+		require.Nil(t, err)
+		defer conn.Close()
+
+		// The plugin handler should be able to use any listed service API version.
+		v1beta1Client := v1beta1.NewExampleClient(conn)
+		v1beta2Client := v1beta2.NewExampleClient(conn)
+
+		// Tests v1beta1 GetExampleInfo
+		_, err = v1beta1Client.GetExampleInfo(context.Background(), &v1beta1.ExampleRequest{})
+		require.Nil(t, err)
+
+		// Tests v1beta1 GetExampleInfo
+		_, err = v1beta2Client.GetExampleInfo(context.Background(), &v1beta2.ExampleRequest{})
+		ret := testCases[callbackCount].returnErr
+		callbackCount++
+		//require.True(t, true)
+		return ret
+	})
+	err = w.Start()
+	require.NoError(t, err)
+
+	p := NewTestExamplePlugin() //first cbk
+	go func() {
+		p.Serve(socketPath)
+	}()
+	require.True(t, waitForPluginRegistrationStatus(t, p.registrationStatus))
+
+	// Trying to start a plugin service at the same socket path should fail
+	// with "bind: address already in use"
+	err = p.Serve(socketPath)
+	require.NotNil(t, err)
+
+	// grpcServer.Stop() will remove the socket and starting plugin service
+	// at the same path again should succeeds and trigger another callback.
+	p.Stop()
+	p = NewTestExamplePlugin() // 2nd cbk
+	go func() {
+		p.Serve(socketPath)
+	}()
+	require.True(t, waitForPluginRegistrationStatus(t, p.registrationStatus))
+
+	// Starting another plugin with the same name got verification error.
+	p2 := NewTestExamplePlugin() //3rd cbk
+	socketPath2 := socketDir + "/plugin2.sock"
+	go func() {
+		p2.Serve(socketPath2)
+	}()
+	require.False(t, waitForPluginRegistrationStatus(t, p2.registrationStatus))
+
+	// Restarts plugin watcher should traverse the socket directory and issues a
+	// callback for every existing socket.
+	err = w.Stop()
+	require.NoError(t, err)
+	errCh := make(chan error)
+	go func() {
+		errCh <- w.Start()
+	}()
+	require.True(t, waitForPluginRegistrationStatus(t, p.registrationStatus))
+	require.False(t, waitForPluginRegistrationStatus(t, p2.registrationStatus))
+
+	err = <-errCh
+	require.NoError(t, err)
+
+	err = w.Stop()
+	require.NoError(t, err)
+	err = w.Cleanup()
+	require.NoError(t, err)
+}
+
+func waitForPluginRegistrationStatus(t *testing.T, statusCh chan watcherapi.RegistrationStatus) bool {
+	select {
+	case status := <-statusCh:
+		return status.PluginRegistered
+	case <-time.After(5 * time.Second):
+		t.FailNow()
```

Comment:
Nit: t.Fatalf
FailNow doesn't log the line where it failed nor any error message. It only marks the test as failed which makes it hard to debug